File: reverseVowels.java
import java.util.*;
// very similar to reverse string.java 
// Approach 1- Two Pointers
class Solution { // TC O(n) SC O(n)
    public void swap(char[] sChar, int a, int b){
        char temp= sChar[a];
        sChar[a]=sChar[b];
        sChar[b]=temp;
    }
    
    public String reverseVowels(String s) {
        Set<Character> set= new HashSet<>(Arrays.asList('A','a','E','e',
                                                       'I','i','O','o','U','u'));
        
        int l=0, r=s.length()-1;
        char[] sChar= s.toCharArray();
        while(l<r){
            while(l<s.length()-1 && !set.contains(sChar[l])){
                l++;
            }
            while(r>=0 && !set.contains(sChar[r])){
                r--;
            }
            if (l<r){
                swap(sChar,l++,r--);
            }
        }
        return new String(sChar);
    }
}

class reverseVowels{
    public static void main(String[] args) {
        String x= "hello";
        Solution s= new Solution();
        System.out.println(s.reverseVowels(x));
    }
}

File: reverseVowels.py
# very similar to reverse string.java 
# Approach 1- Two Pointers
def reverseVowels(s): # TC O(n) // SC O(N)
	vowels=['A','a','E','e','I','i','O','o','U','u']
	seen=set(vowels)
	l,r=0,len(s)-1
	s=list(s)
	while l<r:
		while l<=len(s)-1 and s[l] not in seen:
			l+=1
		while r>=0 and s[r] not in seen:
			r-=1
		if (l<r):
			temp=s[l]
			s[l]=s[r]
			s[r]=temp
			l+=1
			r-=1

	return "".join(s)

print(reverseVowels("hello")) 
print(reverseVowels("leetcode")) 
